plugin/keyring_vault/vault_io.cc

```diff
@@ -0,0 +1,136 @@
+#include <my_global.h>
+#include "vault_io.h"
+#include "vault_keys_list.h"
+#include "vault_credentials_parser.h"
+#include <curl/curl.h>
+#include <sstream>
+
+namespace keyring {
+
+my_bool Vault_io::init(std::string *keyring_storage_url)
+{
+  Vault_credentials_parser vault_credentials_parser(logger);
+  Vault_credentials vault_credentials;
+  return vault_credentials_parser.parse(keyring_storage_url, &vault_credentials) ||
+         vault_curl->init(&vault_credentials);
+}
+
+Vault_io::~Vault_io()
+{
+  if(vault_curl != NULL)
+    delete vault_curl;
+  if(vault_parser != NULL)
+    delete vault_parser;
+}
+
+std::string Vault_io::get_errors_from_response(std::string *json_response)
+{
+  if (json_response->empty())
+    return "";
+
+  std::string errors_from_response, err_msg;
+  if(vault_parser->parse_errors(json_response, &errors_from_response))
+    err_msg = " Error while parsing error messages";
+  else if (errors_from_response.size()) //we found error in response
+    err_msg = " Vault has returned the following error(s): " +
+              errors_from_response;
+  return err_msg;
+}
+
+my_bool Vault_io::get_serialized_object(ISerialized_object **serialized_object)
+{
+  static std::string err_msg("Could not retrieve list of keys from Vault.");
+  *serialized_object= NULL;
+  std::string json_response;
+
+  if(vault_curl->list_keys(&json_response))
+  {
+    logger->log(MY_ERROR_LEVEL, (err_msg +
+                get_errors_from_response(&json_response)).c_str());
+    return TRUE;
+  }
+  if (json_response.empty()) //no keys
+  {
+    *serialized_object = NULL;
+    return FALSE;
+  }
+
+  Vault_keys_list *keys = new Vault_keys_list();
```

Comment:
Changed to std::auto_ptr - boost is forbidden in MySQL Oracle and they take it our from the source code.